server/matcher/match_test.go

```diff
@@ -0,0 +1,1444 @@
+			wantNumInserted: 0,
+		},
+		{
+			name: "market buy, 2 maker match",
+			args: args{
+				book:  newBooker(),
+				queue: []order.Order{takers[1]},
+			},
+			doesMatch: true,
+			wantMatches: []*order.Match{
+				newMatch(takers[1], []*order.LimitOrder{bookSellOrders[nSell-1], bookSellOrders[nSell-2]}, 1*LotSize),
+			},
+			wantNumPassed:   1,
+			wantNumFailed:   0,
+			wantNumPartial:  0,
+			wantNumInserted: 0,
+		},
+		{
+			name: "market buy, 3 maker match",
+			args: args{
+				book:  newBooker(),
+				queue: []order.Order{takers[2]},
+			},
+			doesMatch: true,
+			wantMatches: []*order.Match{
+				newMatch(takers[2], []*order.LimitOrder{bookSellOrders[nSell-1], bookSellOrders[nSell-2]}),
+			},
+			wantNumPassed:   1,
+			wantNumFailed:   0,
+			wantNumPartial:  0,
+			wantNumInserted: 0,
+		},
+		{
+			name: "market buy, 99 maker match",
+			args: args{
+				book:  newBooker(),
+				queue: []order.Order{takers[3]},
+			},
+			doesMatch: true,
+			wantMatches: []*order.Match{
+				newMatch(takers[3], bookSellOrdersReverse),
+			},
+			wantNumPassed:   1,
+			wantNumFailed:   0,
+			wantNumPartial:  0,
+			wantNumInserted: 0,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// Reset Filled amounts of all pre-defined orders before each test.
+			resetTakers()
+			resetMakers()
+
+			fmt.Printf("%v\n", takers)
+
+			matches, passed, failed, partial, inserted := me.Match(tt.args.book, tt.args.queue)
+			matchMade := len(matches) > 0 && matches[0] != nil
+			if tt.doesMatch != matchMade {
+				t.Errorf("Match expected = %v, got = %v", tt.doesMatch, matchMade)
+			}
+			if len(matches) != len(tt.wantMatches) {
+				t.Errorf("number of matches %d, expected %d", len(matches), len(tt.wantMatches))
+			}
+			for i := range matches {
+				if !reflect.DeepEqual(matches[i], tt.wantMatches[i]) {
+					t.Errorf("matches[%d] = %v, want %v", i, matches[i], tt.wantMatches[i])
+				}
+			}
+			if len(passed) != tt.wantNumPassed {
+				t.Errorf("number passed %d, expected %d", len(passed), tt.wantNumPassed)
+			}
+			if len(failed) != tt.wantNumFailed {
+				t.Errorf("number failed %d, expected %d", len(failed), tt.wantNumFailed)
+			}
+			if len(partial) != tt.wantNumPartial {
+				t.Errorf("number partial %d, expected %d", len(partial), tt.wantNumPartial)
+			}
+			if len(inserted) != tt.wantNumInserted {
+				t.Errorf("number inserted %d, expected %d", len(inserted), tt.wantNumInserted)
+			}
+		})
+	}
+}
+func Test_shuffleQueue(t *testing.T) {
+	// Setup the match package's logger.
+	startLogger()
+
+	// order queues to be shuffled
+	q3_1 := []order.Order{
+		limitOrders[0],
+		marketOrders[0],
+		marketOrders[1],
+	}
+
+	// q3_2 has same orders as q1 in different order
+	q3_2 := []order.Order{
+		marketOrders[0],
+		limitOrders[0],
+		marketOrders[1],
+	}
+
+	// q3Shuffled is the expected result of sorting q3_1 and q3_2
+	q3Shuffled := []order.Order{
+		limitOrders[0],
+		marketOrders[0],
+		marketOrders[1],
+	}
```

Comment:
Just an observation, but `q3_1` is the same as `q3Shuffled`.

Reply:
Yup. Although if any bit of any of the orders changed, q3Shuffled could end up as neither q3_1 or q3_2, so I duplicated the slice explicitly.